### PR TITLE
Benefits of TDD

### PR DESCRIPTION
**Pull request by Raniz85**

### CODE REVIEW

File: src/posts/benefits-of-tdd/index.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+but I think that the tests supplement any external documentation in a good way which leaves external documentation free to focus on higher-levels and leave out implementation details.
+
+## Confident Refactoring
+With high test coverage we can be confident that whatever changes we make are safe as long as all the tests pass when we are done.
+The fact that test-driving code means it is looser coupled also makes it a lot easier to refactor.
+
+This goes hand in hand with KISS and YAGNI because adhering to both require us to be able to make the necessary changes when they _are_ required.
+If we are confident enough in our test coverage there is no need to overcomplicate things,
+because we know that once we _need_ to make it more complex we can lean back against our tests.
+
```
Comment (Raniz85) [inline src/posts/benefits-of-tdd/index.md:122]: Jag är lite fundersam över om jag kanske ska ha med ett avsnitt om nackdelar också

File: src/posts/benefits-of-tdd/index.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+## Improved Functional Quality
+Since all the functionality has been implemented by testing it first we can be reasonably sure that it is also correct.
+This, of course, requires us to write enough tests to assert correct behaviour from the code,
+which in turn requires discipline in adhering to only driving development forwards by writing more tests.
+
+Another benefit of TDD that contributes to this for me is that I have different mindsets when writing test cases and implementing the functionality they require.
+
+When writing tests I am thinking more about inputs, outputs and error-conditions than I am when I am writing implementation code.
+This leads me to better handling of errors and edge-cases than if I were to just implement something without test-driving it.
```
Comment (nizos) [inline src/posts/benefits-of-tdd/index.md:90]: omit "me"?

File: src/posts/benefits-of-tdd/index.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+Your mileage will vary here, but I think that the more tests you write, the better you get at this.
+
+All this _should_ lead to fewer bugs and hotfixes,
+improving our user's trust in us as well as the developers confidence in the software they produce.
+With enough confidence we might actually deploy to production on a friday afternoon without risking the weekend.
+
+## Bugs are Missing Tests
+Bugs are no longer mistakes, they are test-cases we did not wrote to drive that specific functionality.
```
Comment (nizos) [inline src/posts/benefits-of-tdd/index.md:98]: "did not wrote" kan skrivas om

File: src/posts/benefits-of-tdd/index.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+By thinking about bugs this way we get an opportunity to consider _why_ we never wrote that test from the beginning.
+It might be because the specification was incomplete,
+or it might be because we were not diligent enough in thinking about possible edge- or error-cases.
+In any case, it is an opportunity to either fix our design process, or for personal growth - or a little bit of both.
+
+## Tests are Documentation
+The tests convey the assumptions we had about the implementation when we wrote it,
+serving as a better form of documentation than external text documents -
+or even text documents in the same repository.
+This is because, unlike text documents, the tests needs to change if the implementation change,
```
Comment (nizos) [inline src/posts/benefits-of-tdd/index.md:109]: tests need
implementation changes

File: src/posts/benefits-of-tdd/index.md (hunk 1 above)
Comment (nizos) [reply src/posts/benefits-of-tdd/index.md:122]: Kanske en separat blog post? :)
Typ utmaningar och strategier? :)
Vilka nackdelar ser du?

Comment (Raniz85) [reply src/posts/benefits-of-tdd/index.md:122]: Det är väl mest svårigheter som man behöver lära sig att jobba runt/bli bättre på att hantera.

Men ibland är det helt enkelt inte värt att testdriva nånting för att det är krångligt att skriva testen och antingen trivialt att testa manuellt eller används så pass mycket så att regressioner skulle bli uppenbara direkt.

Comment (Raniz85) [reply src/posts/benefits-of-tdd/index.md:122]: Jag tror att det kan bli lite för kort för en egen bloggpost

File: src/posts/benefits-of-tdd/index.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+By thinking about bugs this way we get an opportunity to consider _why_ we never wrote that test from the beginning.
+It might be because the specification was incomplete,
+or it might be because we were not diligent enough in thinking about possible edge- or error-cases.
+In any case, it is an opportunity to either fix our design process, or for personal growth - or a little bit of both.
+
+## Tests are Documentation
+The tests convey the assumptions we had about the implementation when we wrote it,
+serving as a better form of documentation than external text documents -
+or even text documents in the same repository.
+This is because, unlike text documents, the tests need to change if the implementation changes,
```
Comment (provegard) [inline src/posts/benefits-of-tdd/index.md:109]: Inte nödvändigtvis, testerna kan ju vara relativt frikopplade från implementationen. Men det var kanske inte det du tänkte på här?

Comment (Raniz85) [reply src/posts/benefits-of-tdd/index.md:109]: Nä, precis, det är klumpigt formulerat

Comment (Raniz85) [reply src/posts/benefits-of-tdd/index.md]: Bytte ut _implementation_ mot _behaviour_
